alf/algorithms/hypernetwork_algorithm.py

```diff
@@ -0,0 +1,430 @@
+                   entropy_regularization=None,
+                   state=None):
+        """Perform one batch of training computation.
+
+        Args:
+            inputs (nested Tensor): input training data. 
+            num_particles (int): number of sampled particles. 
+            state: not used
+
+        Returns:
+            AlgorithmStep:
+                outputs: Tensor with shape (batch_size, dim)
+                info: LossInfo
+        """
+        params = self.sample_parameters(num_particles=num_particles)
+        if entropy_regularization is None:
+            entropy_regularization = self._entropy_regularization
+
+        return self._generator.train_step(
+            inputs=None,
+            loss_func=functools.partial(neglogprob, inputs, self._param_net,
+                                        self._loss_type),
+            outputs=params,
+            entropy_regularization=entropy_regularization,
+            state=())
+
+    def evaluate(self, num_particles=None):
+        """Evaluate on a randomly drawn network. """
+
+        assert self._test_loader is not None, "Must set test_loader first."
+        logging.info("==> Begin testing")
+        if self._use_fc_bn:
+            self._generator.eval()
+        params = self.sample_parameters(num_particles=num_particles)
+        self._param_net.set_parameters(params)
+        if self._use_fc_bn:
+            self._generator.train()
+        with record_time("time/test"):
+            if self._loss_type == 'classification':
+                test_acc = 0.
+            test_loss = 0.
+            for i, (data, target) in enumerate(self._test_loader):
+                data = data.to(alf.get_default_device())
+                target = target.to(alf.get_default_device())
+                output, _ = self._param_net(data)  # [B, N, D]
+                loss, extra = self._vote(output, target)
+                if self._loss_type == 'classification':
+                    test_acc += extra.item()
+                test_loss += loss.loss.item()
+
+        if self._loss_type == 'classification':
+            test_acc /= len(self._test_loader.dataset)
+            alf.summary.scalar(name='eval/test_acc', data=test_acc * 100)
+        if self._logging_evaluate:
+            if self._loss_type == 'classification':
+                logging.info("Test acc: {}".format(test_acc * 100))
+            logging.info("Test loss: {}".format(test_loss))
+        alf.summary.scalar(name='eval/test_loss', data=test_loss)
+
+    def _classification_vote(self, output, target):
+        """ensmeble the ooutputs from sampled classifiers."""
+        num_particles = output.shape[1]
+        probs = F.softmax(output, dim=-1)  # [B, N, D]
+        if self._voting == 'soft':
+            pred = probs.mean(1).cpu()  # [B, D]
+            vote = pred.argmax(-1)
+        elif self._voting == 'hard':
+            pred = probs.argmax(-1).cpu()  # [B, N, 1]
+            vote = pred.mode(1)[0]  # [B, 1]
```

Comment:
torch.mode returns only the first index if there are multiple indices with the same number of "votes". A better option could be to uniformly sample one of the indices for the prediction if there are multiple modes.

Reply:
To elaborate:
compute a tensor of real valued predictions `pred`
items, indices = pred.unique(return_counts=True)
modes = torch.nonzero(indices==indices.max()).view(-1)  # modes holds a tensor of the most occurring values
idx = modes[torch.randint(len(modes), (1,))} # get a random mode index
rand_mode = modes[idx]  # output mode value

Reply:
I have not figured out a way to do this in batch mode without converting back to numpy, so I used a for loop.